File: cpp/leetcode/079.cpp
class Solution {
public:
    bool exist(vector<vector<char> > &board, string word) {
        if (word.empty()) return true;
        if (board.empty() || board[0].empty()) return false;
        int m = board.size(), n = board[0].size();
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (search(board, word, 0, i, j, m, n)) return true;
            }
        }
        return false;
    }
    bool search(vector<vector<char> > &board, string word, int idx, int i, int j, int m, int n) {
        if (idx == word.size()) return true;
        if (i < 0 || j < 0 || i >= m || j >= n || board[i][j] != word[idx]) return false;
        char tmp = board[i][j];
        board[i][j] = '#';
        bool res = search(board, word, idx + 1, i - 1, j, m, n) 
                 || search(board, word, idx + 1, i + 1, j, m, n)
                 || search(board, word, idx + 1, i, j - 1, m, n)
                 || search(board, word, idx + 1, i, j + 1, m, n);
        board[i][j] = tmp;
        return res;
    }
};

File: python/leetcode/079.py
class Solution:
    def exist(self, board, word):
        if not word: return True
        if not board or not board[0]: return False
        m, n = len(board), len(board[0])
        for i in xrange(m):
            for j in xrange(n):
                if self.dfs(board, i, j, m, n, 0, word): return True
        return False

    def dfs(self, board, i, j, m, n, idx, word):
        if len(word) == idx: return True
        if i < 0 or j < 0 or i >= m or j >= n or board[i][j] != word[idx]: return False 
        ch, board[i][j] = board[i][j], '#'
        res = self.dfs(board, i - 1, j, m, n, idx + 1, word) or \
              self.dfs(board, i + 1, j, m, n, idx + 1, word) or \
              self.dfs(board, i, j - 1, m, n, idx + 1, word) or \
              self.dfs(board, i, j + 1, m, n, idx + 1, word) 
        board[i][j] = ch
        return res